File: AXPY/Python/GeneratedCode.py
# numpy

# Prompt: "Create 3 code suggestions using the following parameters: Python numpy AXPY def."

# VERSION 1 #

import numpy as np

# Define a vector of numbers
a = np.array([1, 2, 3])

# Multiply each element by a scalar value
b = np.array([4, 5, 6])
c = np.axpy(a, b)
print(c) # Output: array([4, 10, 18])


# VERSION 2 #

import numpy as np

# Define two matrices
A = np.array([[1, 2], [3, 4]])
B = np.array([[5, 6], [7, 8]])

# Perform matrix multiplication
C = np.dot(A, B)
print(C) # Output: array([[19, 22], [43, 50]])


# VERSION 3 #

import numpy as np

# Define two arrays with different shapes
a = np.array([1, 2, 3])
b = np.array([4, 5, 6])

# Perform element-wise multiplication
c = np.multiply(a, b)
print(c) # Output: array([4, 10, 18])


# ----------------------------------------------------------------------------------------------------------------------------------------

# cuPy

# Prompt: "Create 3 code suggestions using the following parameters: Python cuPy AXPY def."

# VERSION 1 #

import cupy
import numpy as np

# Define a matrix and a vector
A = cupy.eye(3)
x = cupy.array([1, 2, 3])

# Perform a matrix-vector dot product using AXPY
y = cupy.zeros((3,))
cupy.axpy(A, x, y)

print(y) # Output: array([2., 4., 6.])


# VERSION 2 #

import cupy
import numpy as np

# Define a batch of matrices and vectors
A = cupy.stack((cupy.eye(3), cupy.eye(3)))
x = cupy.stack((cupy.array([1, 2, 3]), cupy.array([4, 5, 6])))

# Perform a batched matrix-vector dot product using AXPY
y = cupy.zeros((2, 3))
cupy.axpy(A, x, y)

print(y) # Output: array([[2., 4., 6.],
                       [7., 10., 13.]])


# VERSION 3 #

import cupy
import numpy as np

# Define a custom matrix-vector dot product function using AXPY
def dot_product(A, x):
    y = cupy.zeros((A.shape[0],))
    cupy.axpy(A, x, y)
    return y

# Test the function
A = cupy.eye(3)
x = cupy.array([1, 2, 3])
y = dot_product(A, x)
print(y) # Output: array([2., 4., 6.])

# ----------------------------------------------------------------------------------------------------------------------------------------

# pyCUDA

# Prompt: "Create 3 code suggestions using the following parameters: Python pyCUDA AXPY def."

# VERSION 1 #

import pycuda.driver as cudadrv
from pycuda.compiler import SourceModule

# Define a kernel for the vector-matrix multiplication
mod = SourceModule("""
__global__ void axpy(float **A, float **x, float **y, int N, int M) {
    int i = blockIdx.x * blockDim.x + threadIdx.x;
    if (i < N) {
        for (int j = 0; j < M; ++j) {
            y[i][j] = A[i][j] * x[i][j];
        }
    }
}
 """)

# Load the kernel into the GPU
dev = cudadrv.Device()
func = mod.get_function("axpy")

# Allocate memory on the GPU for the inputs and outputs
A = cudadrv.to_device(numpy.random.rand(10, 10).astype(numpy.float32))
x = cudadrv.to_device(numpy.random.rand(10, 10).astype(numpy.float32))
y = cudadrv.to_device(numpy.zeros((10, 10)).astype(numpy.float32))

# Set the number of blocks and threads per block
block_size = 16
grid_size = (10 + block_size - 1) // block_size

# Launch the kernel
func(A, x, y, 10, 10, block_size, grid_size)

# Synchronize the threads
cudadrv.synchronize()

# Copy the result back to the CPU
result = y.copy_to_host()
print(result)


# VERSION 2 #

import pycuda.driver as cudadrv
from pycuda.compiler import SourceModule

# Define a kernel for the matrix-vector multiplication
mod = SourceModule("""
__global__ void axpy(float **A, float **x, float **y, int N, int M) {
    int i = blockIdx.x * blockDim.x + threadIdx.x;
    if (i < N) {
        for (int j = 0; j < M; ++j) {
            y[i][j] = A[i][j] * x[j];
        }
    }
}
 """)

# Load the kernel into the GPU
dev = cudadrv.Device()
func = mod.get_function("axpy")

# Allocate memory on the GPU for the inputs and outputs
A = cudadrv.to_device(numpy.random.rand(10, 10).astype(numpy.float32))
x = cudadrv.to_device(numpy.random.rand(10).astype(numpy.float32))
y = cudadrv.to_device(numpy.zeros((10, 10)).astype(numpy.float32))

# Set the number of blocks and threads per block
block_size = 16
grid_size = (10 + block_size - 1) // block_size

# Launch the kernel
func(A, x, y, 10, 10, block_size, grid_size)

# Synchronize the threads
cudadrv.synchronize()

# Copy the result back to the CPU
result = y.copy_to_host()
print(result)


# VERSION 3 #

# ONLY 2 CODES WERE PROVIDED #


# ----------------------------------------------------------------------------------------------------------------------------------------

# Numba

# Prompt: "Create 3 code suggestions using the following parameters: Python Numba AXPY def."

# VERSION 1 #

import numba

@numba.jit
def axpy_loop(a, b, n):
    for i in range(n):
        a[i] += b[i]

a = np.array([1, 2, 3])
b = np.array([4, 5, 6])
n = len(a)
axpy_loop(a, b, n)
print(a) # Output: array([5, 7, 9])


# VERSION 2 #

import numba
import numpy as np
from numba import cuda

@numba.jit
@cuda.jit
def axpy_gpu(a, b, n):
    dev = cuda.device()
    a_dev = cuda.to_device(a, dev)
    b_dev = cuda.to_device(b, dev)
    c_dev = cuda.empty_like(a_dev)
    @cuda.kernel
    def kernel(i, j):
        c_dev[i] = a_dev[i] + b_dev[j]
    cuda.parallel.forall(range(n), kernel, block=128)
    return c_dev.copy_to_host()

a = np.random.rand(1000)
b = np.random.rand(1000)
n = 1000
result = axpy_gpu(a, b, n)
print(result)


# VERSION 3 #

import numba

@numba.jit
def axpy_numba(a, b, n):
    return numba.axpy(a, b, n)

a = np.array([1, 2, 3])
b = np.array([4, 5, 6])
n = len(a)
result = axpy_numba(a, b, n)
print(result) # Output: array([5, 7, 9])

File: AXPY/Julia/GeneratedCode.jl

# ===================================================================================================================
#               Threads Programming Model
# ===================================================================================================================

# Prompt: "Create 3 code suggestions using the following parameters: Julia Threads AXPY."

# SUGGESTION 1 # ----------------------------------

using LinearAlgebra
using ThreadPools

# Define matrix dimensions
n = 1000
m = 1000
k = 1000

# Create matrices
A = rand(Float64, n, k)
B = rand(Float64, k, m)
C = similar(A, n, m)

# Define thread pool
pool = ThreadPool(4) # 4 threads

# Perform matrix multiplication using AXPY
@sync for i in 1:n
    for j in 1:m
        C[i, j] = dot(A[:, i], B[j, :])
end

# Print result
println("Matrix product: $(C)")


# SUGGESTION 2 # ----------------------------------
  
using LinearAlgebra
using ThreadPools

# Define matrix dimensions
n = 1000
p = 100

# Create matrices
A = rand(Float64, n, p)
b = rand(Float64, n)

# Define thread pool
pool = ThreadPool(4) # 4 threads

# Perform linear least squares using AXPY
x = zeros( Float64, p )
@sync for i in 1:n
    for j in 1:p
        x[j] += dot(A[:, i], b[i])
end

# Print solution
println("Least squares solution: $x")


# SUGGESTION 3 # ----------------------------------

using LinearAlgebra
using ThreadPools

# Define matrix dimension
n = 1000

# Create matrix
A =rand(Float64, n, n)

# Define thread pool
pool = ThreadPool(4) # 4 threads

# Perform eigenvalue decomposition using AXPY
λ, V = eigs(A, nev=10)

# Print eigenvalues and corresponding vectors
println("Eigenvalues: $(λ)")
println("Eigenvectors: $(V[:, 1:10])")



# ===================================================================================================================
#               CUDA Programming Model
# ===================================================================================================================

# Prompt: "Create 3 code suggestions using the following parameters: Julia CUDA AXPY."

# SUGGESTION 1 # ----------------------------------

using CUDA
using LinearAlgebra

# Define matrix dimensions
n = 1000
m = 1000
k = 1000

# Create matrices
A = cu(rand(Float64, n, k))
B = cu(rand(Float64, k, m))
C = similar(A, n, m)

# Define thread block size
blockSize = (32, 32)

# Shared memory size
sharedMem = 16 * 1024

# Number of blocks
numBlocks = div(n, blockSize[1])

# Perform matrix multiplication using AXPY
@cuda begin
    @kernel void axpy!(Int(n), Int(m), Int(k), Float64(A), Float64(B), Float64(C), sharedMem, blockSize)
        @inbounds begin
            # Compute row and column indices within block
            i = blockIdx().x * blockSize().x + threadIdx().x
            j = blockIdx().y * blockSize().y + threadIdx().y

            # Compute partial sum for current element
            var partialSum = zero(Float64)
            for l in 1:k
                partialSum += A[i, l] * B[l, j]

            # Store partial sum in shared memory
            shared Mem[threadIdx().x + threadIdx().y * sharedMem] = partialSum

            # Synchronize threads within block
            syncthreads()

            # Update final result
            if threadIdx().x == 0 && threadIdx().y == 0
                C[i, j] = sharedMem[threadIdx().x + threadIdx().y * sharedMem]
            end
        end
    end
end

# Print result
println("Matrix product: $(C)")


# SUGGESTION 2 # ----------------------------------

using CUDA
using LinearAlgebra

# Define matrix dimensions
n = 1000
p = 100

# Create matrices
A = cu(rand(Float64, n, p))
b = cu(rand(Float64, n))

# Define thread block size
blockSize = (32, 32)

# Shared memory size
sharedMem = 16 * 1024

# Number of blocks
numBlocks = div(n, blockSize[1])

# Perform linear least squares using AXPY
@cuda begin
    @kernel void axpy!(Int(n), Int(p), Float64(A), Float64(b), Float64(x), sharedMem, blockSize)
        @inbounds begin
            # Compute row and column indices within block
            i = blockIdx().x * blockSize().x + threadIdx().x
            j = blockIdx().y * blockSize().y + threadIdx().y

            # Compute partial sum for current element
            var partialSum = zero(Float64)
            for l in 1:p
                partialSum += A[i, l] * b[l]

            # Store partial sum in shared memory
            shared Mem[threadIdx().x + threadIdx().y * sharedMem] = partialSum

            # Synchronize threads within block
            syncthreads()

            # Update final result
            if threadIdx().x == 0 && threadIdx().y == 0
                x[j] = sharedMem[threadIdx().x + threadIdx().y * sharedMem]
            end
        end
    end
end

# Print solution
println("Least squares solution: $x")


# SUGGESTION 3 # ----------------------------------

using CUDA
using LinearAlgebra

# Define matrix dimensions
n = 1000
m = 1000
k = 1000

# Create matrices
A = cu(rand(Float64, n, k))
B = cu(rand(Float64, k, m))
C = similar(A, n, m)

# Define thread block size
blockSize = (32, 32)

# Shared memory size
sharedMem = 16 * 1024

# Number of blocks
numBlocks = div(n, blockSize[1])

# Perform matrix-matrix multiplication using AXPY
@cuda begin
    @kernel void axpy!(Int(n), Int(m), Int(k), Float64(A), Float64(B), Float64(C), sharedMem, blockSize)
        @inbounds begin
            # Compute row and column indices within block
            i = blockIdx().x * blockSize().x + threadIdx().x
            j = blockIdx().y * blockSize().y + threadIdx().y

            # Compute partial sum for current element
            var partialSum = zero(Float64)
            for l in 1:k
                partialSum += A[i, l] * B[l, j]

            # Store partial sum in shared memory
            shared Mem[threadIdx().x + threadIdx().y * sharedMem] = partialSum

            # Synchronize threads within block
            syncthreads()

            # Update final result
            if threadIdx().x == 0 && threadIdx().y == 0
                C[i, j] = sharedMem[threadIdx().x + threadIdx().y * sharedMem]
            end
        end
    end
end

# Print result
println("Matrix-matrix product: $(C)")



# ===================================================================================================================
#               AMDGPU Programming Model
# ===================================================================================================================

# Prompt: "Create 3 code suggestions using the following parameters: Julia AMDGPU AXPY."

# SUGGESTION 1 # ----------------------------------

using AMDGPU
using LinearAlgebra

# Define matrix dimensions
n = 1000
m = 1000
k = 1000

# Create matrices
A = rand(Float64, n, k)
B = rand(Float64, k, m)
C = similar(A, n, m)

# Define thread block size
blockSize = (32, 32)

# Shared memory size
sharedMem = 16 * 1024

# Number of blocks
numBlocks = div(n, blockSize[1])

# Perform matrix-matrix multiplication using AXPY
@amdgpu begin
    @kernel void axpy!(Int(n), Int(m), Int(k), Float64(A), Float64(B), Float64(C), sharedMem, blockSize)
        @inbounds begin
            # Compute row and column indices within block
            i = blockIdx().x * blockSize().x + threadIdx().x
            j = blockIdx().y * blockSize().y + threadIdx().y

            # Compute partial sum for current element
            var partialSum = zero(Float64)
            for l in 1:k
                partialSum += A[i, l] * B[l, j]

            # Store partial sum in shared memory
            shared Mem[threadIdx().x + threadIdx().y * sharedMem] = partialSum

            # Synchronize threads within block
            syncthreads()

            # Update final result
            if threadIdx().x == 0 && threadIdx().y == 0
                C[i, j] = sharedMem[threadIdx().x + threadIdx().y * sharedMem]
            end
        end
    end
end

# Print result
println("Matrix-matrix product: $(C)")


# SUGGESTION 2 # ----------------------------------

using AMDGPU
using LinearAlgebra

# Define vector dimensions
n = 1000

# Create vectors
a = rand(Float64, n)
b = rand(Float64, n)

# Define thread block size
blockSize = (32, )

# Shared memory size
sharedMem = 16 * 1024

# Number of blocks
numBlocks = div(n, blockSize[1])

# Perform vector-vector dot product using AXPY
@amdgpu begin
    @kernel void axpy!(Int(n), Float64(a), Float64(b), Float64(dotProduct), sharedMem, blockSize)
        @inbounds begin
            # Compute index within block
            i = blockIdx().x * blockSize().x + threadIdx().x

            # Compute partial sum for current element
            var partialSum = zero(Float64)
            for j in 1:n
                partialSum += a[i] * b[j]

            # Store partial sum in shared memory
            shared Mem[threadIdx().x + threadIdx().y * sharedMem] = partialSum

            # Synchronize threads within block
            syncthreads()

            # Update final result
            if threadIdx().x == 0 && threadIdx().y == 0
                dotProduct[i] = sharedMem[threadIdx().x + threadIdx().y * sharedMem]
            end
        end
    end
end

# Print result
println("Dot product: $(dotProduct)")


# SUGGESTION 3 # ----------------------------------

using AMDGPU

# Define the arrays
a = zeros(Float64, 1000)
b = zeros(Float64, 1000)

# Set the beta parameter
beta = 2.0

# Perform an all-to-all Axpy operation
AXPY!(a, b, beta)

# Print the results
println(a)



# ===================================================================================================================
#               KernelAbstractions Programming Model
# ===================================================================================================================

# Prompt: "Create 3 code suggestions using the following parameters: Julia KernelAbstractions AXPY."

# SUGGESTION 1 # ----------------------------------

using KernelAbstractions

# Define the kernel
kernel = @kernel begin
    # Define the input arrays
    a, b = @in(Array{Float64}, 2)
    
    # Define the output array
    c = similar(a)
    
    # Perform the Axpy operation
    for i in 1:length(a)
        for j in 1:length(b)
            c[i] += a[i] * b[j]
        end
    end
    
    return c
end

# Create some test data
a = rand(Float64, 1000)
b = rand(Float64, 1000)

# Run the kernel
c = run(kernel, (a, b))

# Print the result
println(c)


# SUGGESTION 2 # ----------------------------------

using KernelAbstractions
using LinearAlgebra

# Define the Axpy function
function axpy(a, b, beta)
    # Perform the Axpy operation
    for i in 1:length(a)
        for j in 1:length(b)
            a[i] += beta * b[j]
        end
    end
end

# Wrap the Axpy function with a kernel
kernel = @kernel axpy(a, b, beta)

# Create some test data
a =rand(Float64, 1000)
b = rand(Float64, 1000)
beta = 2.0

# Run the kernel
result = run(kernel, (a, b, beta))

# Print the result
println(result)


# SUGGESTION 3 # ----------------------------------

using KernelAbstractions

# Define the serial Axpy function
function axpy_serial(a, b, beta)
    # Perform the Axpy operation
    for i in 1:length(a)
        for j in 1:length(b)
            a[i] += beta * b[j]
        end
    end
end

# Wrap the serial Axpy function with a kernel
kernel = @kernel axpy_serial(a, b, beta)

# Create some test data
a = rand(Float64, 1000)
b = rand(Float64, 1000)
beta = 2.0

# Run the kernel in parallel
using Threads
nthreads = 4
@sync begin
    for i in 1:nthreads
        run(kernel, (a, b, beta), threads=i)
    end
end

# Print the result
println(a)



# ---------------------------------------------------------------------------------------------------------------------------------------------------------
